### PR TITLE
Introduce threading and some refactoring

### PR DESCRIPTION
**Pull request by derrickchua**

### CODE REVIEW

**Comment by derrickchua** [comment]
DO NOT merge. Getting a java.lang.ArrayIndexOutOfBoundsException: 10
	at tetris.State.makeMove(State.java:205)
	at tetris.State.makeMove(State.java:195)
	at tetris.State.makeMove(State.java:190)
	at tetris.scorer.Scorer.play(Scorer.java:41)
	at tetris.scorer.Scorer.play(Scorer.java:30)
        at tetris.genetic.GeneticAlgorithm$HeuristicRunner.run(GeneticAlgorithm.java:272)

File: src/main/java/tetris/genetic/GeneticAlgorithm.java [hunk 1]
```diff
@@ -209,6 +229,7 @@ private void score() throws Exception {
     executor.shutdown();
     executor.awaitTermination(1000, TimeUnit.MINUTES);
 
+    // Below code is for testing purposes, to be removed
```
Comment (jethrokuan) [inline src/main/java/tetris/genetic/GeneticAlgorithm.java]: can we delete this

File: src/main/java/tetris/genetic/GeneticAlgorithm.java [hunk 2]
```diff
@@ -185,6 +201,10 @@ private void generateNewHeuristics() throws Exception {
 
   }
 
+  /** We thread this for each heuristic, which will each play games as shown in HeuristicRunner
+   *
+   * @throws Exception
+   */
   private void score() throws Exception {
 
 //    ForkJoinPool pool = new ForkJoinPool(POPULATION_SIZE);
```
Comment (jethrokuan) [inline src/main/java/tetris/genetic/GeneticAlgorithm.java]: delete this

File: src/main/java/tetris/scorer/Scorer.java [hunk 3]
```diff
@@ -65,19 +63,23 @@ public double getAverageScore() {
 
 //  public static void main(String[] args) {
 //    ArrayList<Feature> features = new ArrayList<Feature>();
```
Comment (jethrokuan) [inline src/main/java/tetris/scorer/Scorer.java]: delete this too

File: src/main/java/tetris/State.java [hunk 4]
```diff
@@ -202,6 +202,9 @@ public boolean makeMove(int orient, int slot) {
     int height = top[slot]-pBottom[nextPiece][orient][0];
     //for each column beyond the first in the piece
     for(int c = 1; c < pWidth[nextPiece][orient];c++) {
+      if (slot + c >= 10) {
```
Comment (jethrokuan) [inline src/main/java/tetris/State.java]: is this still happening? if not let's remove it

File: src/main/java/tetris/heuristic/Heuristic.java [hunk 5]
```diff
@@ -72,13 +70,18 @@ public Heuristic(File file) {
...
-    this.weights = new Random().doubles(this.size, -1, 1).toArray();
+    Random r = new Random();
+    this.weights = new Double[this.size];
+    for (int i = 0; i < this.size; i++) {
+      this.weights[i] = r.nextDouble();
+    }
   }
 
-  public Heuristic(ArrayList<Feature> features, double[] heuristicArray) {
+  public Heuristic(ArrayList<Feature> features, Double[] heuristicArray, Double score) {
```
Comment (jethrokuan) [inline src/main/java/tetris/heuristic/Heuristic.java]: naise

File: src/main/java/tetris/genetic/GeneticAlgorithm.java [hunk 6]
```diff
@@ -31,18 +31,16 @@
...
   private static final Integer NUM_GENERATIONS = 100;
-  private static final Integer NUM_GAMES = 50;
+  private static final Integer NUM_GAMES = 10;
+  private static final Integer POPULATION_SIZE = 200;
 
   private static final Double MUTATION_RATE = 0.1;
-  private static final Integer SELECTION = 5;
   private static final Double DEFAULT_SCORE = 0.0;
 
   private static final Integer SURVIVORS = 15;
```
Comment (alexanderleegs) [inline src/main/java/tetris/genetic/GeneticAlgorithm.java]: Are we still leaving so many survivors from the last round? Would something like 5 work better?

**Pull request merged**